joomla/joomla-cms
# TinyMCE drag&drop toolbar editing

### Summary of Changes

The pull request add ability to drag&drop toolbar editing to help to configure TinyMCE editor.
## How it works

![screen 2016-09-11 13 50 00 1374x694](https://cloud.githubusercontent.com/assets/1568198/18416811/c9899418-7827-11e6-9e32-567626215c30.png)
1. Available buttons, and menu (Source panel). Which can be used for build your own toolbar.
2. Your toolbar (Target panel). Drop here an available buttons.
3. The tabs of Sets. By default it is 3 set. but it can be unlimited. Number of the sets can be changed in advanced tab:
   ![screen 2016-09-11 13 50 17 573x202](https://cloud.githubusercontent.com/assets/1568198/18416817/292a739c-7828-11e6-9bdf-eaa03f8e8b5c.png)
4. The Builder have 3 preset: Simple, Medium, Advanced, which can be used for quick setup. The Clear button will clear the canvas.
5. Access level for a current Set.
6. The form with additional options for a current Set.
7. Drop zone for Menubar.
8. Drop zone for Toolbar1.
9. Drop zone for Toolbar2.

**How to add button to my toolbar?**
Drag the button from Source panel (1) to Target panel (2), and drop it there.

**How to remove the button?**
Drag&drop unneeded button out of the Target panel (2).

**I want more Sets!!!**
Set the amount of sets in the advanced tab. A new sets will appears after save.
### Testing Instructions

~~_it does not ready for final test_~~

First of all make sure if the drag&drop editing works.
Then need to check that all existing features still works, as here is huge code changes.
### Documentation Changes Required

Update info about how to configure TinyMCE plugin (if there any).
### B. C.

It can break half of user settings for TinyMCE plugin. So need to test each case: new installation, update.

After upgrade or on clear installation the editor will use next default presets:
- _Advanced_ => for Special access level
- _Medium_ => for Registered
- _Simple_ => for everyone else
### ~~Help needed~~

~~I made 3 presets: Simple, Medium, Advanced.~~

~~Would be cool if someone will improve them, and make it more real for what user need.~~
~~Also would be cool if they will be similar to "old toolbar", so we will keep B/C a little bit~~ 😉 
~~It is easy to edit, just add needed buttons in: https://github.com/Fedik/joomla-cms/blob/0b4d803c35c3995934b017f74496d0c3c26f8bb0/plugins/editors/tinymce/tinymce.php#L1064-L1106~~

**P.S.**
~~I found that make editing depend from "view level" is not as good as I thought on start~~ 😄 
~~So I have an idea to change it a bit. And make a "Presets" so user can set access level for each "Preset". Also it will allow to use separated "Preset" for each user~~ It is already implemented 😉  

I would like to hear the feedback 😉 

~~The pull depending from #11157. But can be tested stand alone.~~
